server/src/main/java/org/elasticsearch/indices/recovery/MultiFileWriter.java

```diff
@@ -39,10 +41,12 @@
 import java.util.Map;
 import java.util.PriorityQueue;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
 
-public class MultiFileWriter implements Releasable {
+public class MultiFileWriter extends AbstractRefCounted implements Releasable {
```

Comment:
I don't understand this change. In particular it looks like you're not explicitly calling `incRef()` anywhere. Given that the `AtomicBoolean closed` is used to guard against double-closing, why make this subclass AbstractRefCounted here?

Reply:
We explicitly incRef [here](https://github.com/dnhatn/elasticsearch/blob/24025c83d071d5c7629e74fae5be5ef4484f65c1/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java#L522-L523). We have to do this because writing file chunks can happen concurrently with close. If MultiFileTransfer hits an error, it will close the MultiFileWriter; and if chunk responses arrive afterward, we will open some files but never close them.

Guarding against double-closing is not required. We can remove it.

Reply:
Sorry I missed the incRef call. Let's leave the double-closed protection in.